packages/https-proxy/lib/server.coffee

```diff
@@ -69,42 +80,90 @@ class Server
       res.end()
     .pipe(res)
 
+  _upstreamProxyForHostPort: (hostname, port) ->
+    getProxyForUrl("https://#{hostname}:#{port}")
+
   _makeDirectConnection: (req, socket, head) ->
     { port, hostname } = url.parse("http://#{req.url}")
 
+    if upstreamProxy = @_upstreamProxyForHostPort(hostname, port)
+      return @_makeUpstreamProxyConnection(upstreamProxy, socket, head, port, hostname)
+
+    debug("Making direct connection to #{hostname}:#{port}")
     @_makeConnection(socket, head, port, hostname)
 
   _makeConnection: (socket, head, port, hostname) ->
-    cb = ->
+    onConnect = ->
       socket.pipe(conn)
       conn.pipe(socket)
       socket.emit("data", head)
 
       socket.resume()
 
-    ## compact out hostname when undefined
-    args = _.compact([port, hostname, cb])
+    conn = new net.Socket()
+    conn.setNoDelay(true)
 
-    conn = net.connect.apply(net, args)
+    socket.on "close", =>
+      conn.destroy()
 
     conn.on "error", (err) =>
       if @_onError
         @_onError(err, socket, head, port)
 
+    ## compact out hostname when undefined
+    args = _.compact([port, hostname, onConnect])
+    conn.connect.apply(conn, args)
+
+  # todo: as soon as all requests are intercepted, this can go away since this is just for pass-through
+  _makeUpstreamProxyConnection: (upstreamProxy, socket, head, toPort, toHostname) ->
+    debug("making proxied connection to #{toHostname}:#{toPort} with upstream #{upstreamProxy}")
+
+    onUpstreamSock = (err, upstreamSock) ->
+      if @_onError
+        if err
+          return @_onError(err, socket, head, port)
+        upstreamSock.on "error", (err) =>
+          @_onError(err, socket, head, port)
+
+      if not upstreamSock
+        ## couldn't establish a proxy connection, fail gracefully
+        socket.resume()
+        return socket.destroy()
+
+      upstreamSock.setNoDelay(true)
+      upstreamSock.pipe(socket)
+      socket.pipe(upstreamSock)
+      socket.emit("data", head)
```

Comment:
why do we emit `data` here instead of writing to the socket?

Reply:
No reason except for that's how it was done in `_makeConnection`. Fixed, changed this to `proxySocket.write(head)` since it's clearer that way what the actual effect of this is, and updated the other instance in `_makeConnection`.